perf(ngcc): only compute basePaths in TargetedEntryPointFinder when n…

petebacondarwin opened the pull request:
…eeded

Previously the `basePaths` were computed when the finder was instantiated.
This was a waste of effort in the case that the targeted entry-point is already
processed.

This change makes the computation of `basePaths` lazy, so that the work is
only done if they are actually needed.

Fixes #36874

petebacondarwin committed:
perf(ngcc): only compute basePaths in TargetedEntryPointFinder when needed

Previously the `basePaths` were computed when the finder was instantiated.
This was a waste of effort in the case that the targeted entry-point is already
processed.

This change makes the computation of `basePaths` lazy, so that the work is
only done if they are actually needed.

Fixes #36874

JoostK commented:
If I'm not mistaken, wouldn't  the `TargetedEntryPointFinder.targetNeedsProcessingOrCleaning` call to determine if early bailout is possible still need to compute the base paths, through its call to  `getEntryPoint -> computePackagePath` which then uses `basePaths`.

petebacondarwin commented:
Agh!

petebacondarwin commented:
Hmm. So we have two options, I think:

1) Consider the `sourceDirectory` as a basePath only initially when calling `targetNeedsProcessingOrCleaning` and only trigger `getBasePaths` if the package cannot be found without it.
  This would speed up most runs, since most entry-points are in the `node_modules` folder (i.e. the `sourceDirectory`) and only impact on pre-processed entry-points that are to be found via the `paths` mappings anyway.
2) Somehow cache the computed `basePaths` (perhaps in the entry-point manifest file. So that it is only computed once.

I think I lean towards 2). WDYT?

gkalpak reviewed (dismissed):
✨

JoostK commented:
A quick look into `targetNeedsProcessingOrCleaning` reveals that it's mostly interested in the package.json file, which is trivial to load without base paths. Unfortunately, the `compiledByAngular` flag is more tricky, as is may depend on configuration which may be loaded from a package configuration, thus requiring proper package path resolution to know where to look for the package's config file. Would that rule out option 1?

Another option would be to revisit the algorithmic complexity of `getBasePaths` so that it would be fast even when there's over 100 paths, assuming that that is why it's so slow?

gkalpak commented:
FWIW, I would towards option 1 (esp. given that `TargetEntryPointFinder#basePaths` is only used in one place). This will always be faster than optimizing `getBasePaths()` :grin:

If we wanted to get fancy, we could change `TargetEntryPointFinder#basePaths` to a custom structure that has an iterator which return the base path first and then lazily computes the rest of the base paths based on `pathMappings` if needed 🤓

petebacondarwin committed:
fixup! perf(ngcc): speed up the `getBasePaths()` computation

petebacondarwin committed:
fixup! perf(ngcc): speed up the `getBasePaths()` computation

petebacondarwin committed:
fixup! perf(ngcc): speed up the `getBasePaths()` computation

petebacondarwin committed:
fixup! perf(ngcc): only compute basePaths in TargetedEntryPointFinder when needed

petebacondarwin committed:
perf(ngcc): speed up the `getBasePaths()` computation

This function needs to deduplicate the paths that are found from the
paths mappings. Previously this deduplication was not linear and also
called the expensive `relative()` function many times.

This commit, suggested by @JoostK, reduces the complexity of the deduplication
by using a tree structure built from the segments of each path.

petebacondarwin commented:
I've gone with option 1 for now. Let's see how this pans out for @n9niwas's issue.

n9niwas commented:
@petebacondarwin I applied this change and now it takes 50ms (vs 500ms previously) on average to process each module, so much better, thanks!

but I noticed that some modules still take 500ms
some of our tsconfig `paths` are also published through `ng-packagr` for internal use in other repositories
they don't exist in `node_modules` folder and shouldn't be processed by ngcc, but because they have `package.json` [this logic](https://github.com/angular/angular-cli/blob/master/packages/ngtools/webpack/src/ngcc_processor.ts#L168) still sends them to ngcc

not sure it's in the scope of this PR, but it would be nice to ignore those too

petebacondarwin commented:
If you have packages imported via `paths mappings` then we still need to check them via ngcc, and unfortunately we have to compute the package folder of these by first computing the base paths. We could possibly get further performance improvements by caching these base paths once they are computed.

gkalpak reviewed (approved):
Missed opportunity to go fancy with iterators and generators, but works for me 😄

JoostK commented on packages/compiler-cli/ngcc/src/entry_point_finder/utils.ts:
```diff
@@ -118,3 +94,76 @@ export function trackDuration<T = void>(task: () => T extends Promise<unknown>?
+    // We hit a leaf so don't bother processing any more of the path
+    return;
+  }
+  const next = segments.shift()!;
```
```suggestion
  const next = segments.shift();
```

JoostK commented on packages/compiler-cli/ngcc/src/entry_point_finder/utils.ts:
```diff
@@ -118,3 +94,76 @@ export function trackDuration<T = void>(task: () => T extends Promise<unknown>?
+function flattenTree(node: Node, paths: AbsoluteFsPath[]) {
+  if (node.children === undefined) {
+    // We found a leaf so store the currentPath
+    paths.push(absoluteFrom(node.path));
```
```suggestion
    paths.push(node.path);
```

JoostK commented on packages/compiler-cli/ngcc/src/entry_point_finder/utils.ts:
```diff
@@ -118,3 +94,76 @@ export function trackDuration<T = void>(task: () => T extends Promise<unknown>?
+ * one of the base paths.)
+ *
+ */
+function dedupePaths(paths: AbsoluteFsPath[]) {
```
I would add an explicit return type here:

```suggestion
function dedupePaths(paths: AbsoluteFsPath[]): AbsoluteFsPath[] {
```

JoostK commented on packages/compiler-cli/ngcc/src/entry_point_finder/utils.ts:
```diff
@@ -118,3 +94,76 @@ export function trackDuration<T = void>(task: () => T extends Promise<unknown>?
+ * Given `['a/b/c', 'a/b/x', 'a/b', 'd/e', 'd/f']` we will end up with `['a/b', 'd/e', 'd/f]`.
+ * (Note that we do not get `d` even though `d/e` and `d/f` share a base directory, since `d` is not
+ * one of the base paths.)
+ *
```
```suggestion

```

JoostK reviewed (approved):
What a lovely algorithm you have there 😄

I did some tests when there's lots of path mappings and the algorithmic improvement to the filtering of contained paths makes a significant difference.

Tests results for 400 invocations of `getBasePaths` when there's 300 path mappings:

**Before**
<img width="422" alt="Screenshot 2020-05-02 at 00 38 20" src="https://user-images.githubusercontent.com/123679/80846673-52cf4400-8c0d-11ea-987d-b624135f3d82.png">

**After** 🎉 
<img width="414" alt="Screenshot 2020-05-02 at 00 38 40" src="https://user-images.githubusercontent.com/123679/80846674-5367da80-8c0d-11ea-8262-3c2a34a8859e.png">

petebacondarwin commented:
@JoostK I added a few more improvements: ditched the recursion, switched from `Map()` to records, minimized the mutations to arrays and object.

petebacondarwin commented:
I also did some benchmarking... From randomized paths I found that the average time "per original path" was reduced to about 20% from the original algorithm.

JoostK commented:
I am seeing some interesting perf characteristics after the latest fixup: `flattenTree` is _a lot_ slower when using records, presumably because `Object.values` is not nearly as fast as iterating a `Map`'s values. The perf of `addPaths` is also slightly better for `Map`. Other than that, the original, recursive variant of `addPath` is surprisingly faster (~35%) for unknown reason.

I would definitely switch back to using `Map`, other than that LGTM.

petebacondarwin commented:
Are you saying the recursive version is faster? Should we switch back to that version? What are you using to benchmark?

petebacondarwin commented:
When I benchmarked again today running the following code 100x on each algorithm:

```
    function benchmark(fn: typeof dedupePaths) {
      const paths = pathStrings.map(absoluteFrom);
      let deduped = 0;
      const now = milliseconds(process.hrtime());
      for (let i = 0; i < 200; i++) {
        deduped = dedupePaths(paths).length;
      }
      const time = milliseconds(process.hrtime()) - now;
      console.log([
        fn.name, os, `${paths.length}`, `${deduped}`, time, time / paths.length, time / deduped
      ].join('::'));
    }
```

I got

Algorithm | ms per path | ms per deduped path
----------|-------------|-----------------
Map | 2.415962675 |  2.461352285
Record | 2.31117757 |  2.355814758

JoostK commented:
Interesting :-)

Here's my results:

**Map**
```
dedupePaths :: OS/X :: 300 :: 200 :: 96.34445099532604 :: 0.32114816998442014 :: 0.48172225497663024
dedupePaths :: Windows :: 300 :: 200 :: 94.02359199523926 :: 0.3134119733174642 :: 0.47011795997619626
dedupePaths :: Unix :: 300 :: 200 :: 91.67663399875164 :: 0.3055887799958388 :: 0.4583831699937582
dedupePaths :: Native :: 300 :: 200 :: 90.23406199365854 :: 0.30078020664552846 :: 0.4511703099682927
```

**Record**
```
dedupePaths :: OS/X :: 300 :: 200 :: 150.2178319990635 :: 0.500726106663545 :: 0.7510891599953174
dedupePaths :: Windows :: 300 :: 200 :: 132.13881799578667 :: 0.4404627266526222 :: 0.6606940899789333
dedupePaths :: Unix :: 300 :: 200 :: 130.1959529966116 :: 0.4339865099887053 :: 0.650979764983058
dedupePaths :: Native :: 300 :: 200 :: 136.70849999785423 :: 0.45569499999284746 :: 0.6835424999892712
```

This is run from a single test in the `jasmine_node_test` `//packages/compiler-cli/ngcc/test:test` target.

petebacondarwin commented:
I guess you are on Windows? I wonder if there is some difference there?
I ran this 100x and took the average. There were definitely some big outliers, do it might be worth running a few times.

petebacondarwin commented:
Oh also the only difference between the two algorithms was the use of Map vs Record, right?

JoostK commented:
Here's my tweaked setup: https://github.com/JoostK/perf-36881

```
yarn test 10000
```

On my MacBook Pro (late 2013):
```
dedupePathsMap :: 300 :: 200 :: 4819.888596996665 :: 0.0016066295323322215 :: 0.0024099442984983326
dedupePathsRecord :: 300 :: 200 :: 6899.658298999071 :: 0.002299886099666357 :: 0.0034498291494995358
-30.14%
```

I'm using a fairly large iteration count as it shows far more stable results. Using a small iteration count, I'm seeing noticeable differences when e.g. swapping the benchmark ordering.

Anyway, this is more out of curiosity than it actually making a difference in real life.

petebacondarwin commented:
Oh well the differences are minimal. I'll leave it as-is.

gkalpak commented on packages/compiler-cli/ngcc/src/entry_point_finder/utils.ts:
```diff
@@ -118,3 +94,76 @@ export function trackDuration<T = void>(task: () => T extends Promise<unknown>?
+  for (const path of paths) {
+    addPath(root, path);
+  }
+  return flattenTree(root).sort().reverse();
```
Nit: I don't think we need this sorting/reversing now.
(Previously, it was used for identifying contained paths.)

petebacondarwin replied:
If we remove this then unit tests fail... I'll check whether that is just fragile tests or something more significant.

petebacondarwin replied:
This line relies upon the ordering: https://github.com/angular/angular/blob/ca839a0c93d3d20efe02f5a9935d2f7caf519785/packages/compiler-cli/ngcc/src/entry_point_finder/targeted_entry_point_finder.ts#L129-L132

petebacondarwin replied:
Or maybe not...

petebacondarwin replied:
Brittle tests. Fixed and remove this.
